src/libraries/System.Private.Runtime.InteropServices.JavaScript/src/System/Runtime/InteropServices/JavaScript/Runtime.cs

```diff
@@ -35,111 +33,119 @@ public static string InvokeJS(string str)
+            Interop.Runtime.DumpAotProfileData(ref buf, len, extraArg);
         }
 
-        public static object GetGlobalObject(string? str = null)
+        public static int GetCsOwnedObjectJsHandle(JSObject target, int shouldAddInflight)
         {
-            return Interop.Runtime.GetGlobalObject(str);
+            if (shouldAddInflight != 0)
+            {
+                target.AddInFlight();
+            }
+            return target.JSHandle;
         }
 
-        public static void DumpAotProfileData (ref byte buf, int len, string extraArg)
+        internal static bool ReleaseCsOwnedObject(JSObject objToRelease)
         {
-            Interop.Runtime.DumpAotProfileData(ref buf, len, extraArg);
+            lock (_csOwnedObjects)
+            {
+                _csOwnedObjects.Remove(objToRelease.JSHandle);
+                Interop.Runtime.ReleaseCsOwnedObject(objToRelease.JSHandle);
+            }
+            return true;
         }
 
-        public static int BindJSObject(int jsHandle, bool ownsHandle, int mappedType)
+        internal static IntPtr CreateCsOwnedObject(JSObject proxy, string typeName, params object[] parms)
+        {
+            object res = Interop.Runtime.CreateCsOwnedObject(typeName, parms, out int exception);
+            if (exception != 0)
+                throw new JSException((string)res);
+
+            var jsHandle = (int)res;
+
+            lock (_csOwnedObjects)
+            {
+                _csOwnedObjects[jsHandle] = new WeakReference<JSObject>(proxy, trackResurrection: true);
+            }
+
+            return (IntPtr)jsHandle;
+        }
+
+        public static JSObject CreateCSOwnedObject(int jsHandle, int mappedType)
         {
             JSObject? target = null;
 
-            lock (_boundObjects)
+            lock (_csOwnedObjects)
             {
-                if (!_boundObjects.TryGetValue(jsHandle, out WeakReference<JSObject>? reference) ||
+                if (!_csOwnedObjects.TryGetValue(jsHandle, out WeakReference<JSObject>? reference) ||
                     !reference.TryGetTarget(out target) ||
                     target.IsDisposed)
                 {
-                    IntPtr jsIntPtr = (IntPtr)jsHandle;
-                    target = mappedType > 0 ? BindJSType(jsIntPtr, ownsHandle, mappedType) : new JSObject(jsIntPtr, ownsHandle);
-                    _boundObjects[jsHandle] = new WeakReference<JSObject>(target, trackResurrection: true);
+                    target = mappedType > 0 ? BindJSType((IntPtr)jsHandle, mappedType) : new JSObject((IntPtr)jsHandle);
+                    _csOwnedObjects[jsHandle] = new WeakReference<JSObject>(target, trackResurrection: true);
                 }
             }
 
             target.AddInFlight();
 
-            return target.GCHandleValue;
+            return target;
         }
 
-        public static int BindCoreCLRObject(int jsHandle, int gcHandle)
+        public static int TryGetCsOwnedObjectJsHandle(object rawObj)
```

Comment:
I think this should probably return a success bool and have an outparam for the jshandle, since theoretically the JSObject you pass in could have an invalid JSHandle (i.e. it was freed)

Reply:
The usage is like this `_try_get_cs_owned_object_js_handle (root.value)`, so we already have the rooted reference, which should not fizzle out in the process.  If somebody gave us JSObject which is already disposed, that would mean we would not find it in the `mono_wasm_object_registry` by js_handle anyway.

Reply:
Sounds fine then